model/diffusion/diffusion_ibrl.py

```diff
@@ -0,0 +1,286 @@
+            dim=0
+        ).values  # unlike RLPD, IBRL uses the min Q value for actor update
+        loss_actor = -torch.mean(current_q)
+        return loss_actor
+
+    def update_target_critic(self, tau):
+        """need to use ensemble_params instead of critic_networks"""
+        for target_ind, target_critic in enumerate(self.target_networks):
+            for target_param_name, target_param in target_critic.named_parameters():
+                source_param = self.ensemble_params[target_param_name][target_ind]
+                target_param.data.copy_(
+                    target_param.data * (1.0 - tau) + source_param.data * tau
+                )
+
+    def update_target_actor(self, tau):
+        for target_param, source_param in zip(
+            self.target_actor.parameters(), self.network.parameters()
+        ):
+            target_param.data.copy_(
+                target_param.data * (1.0 - tau) + source_param.data * tau
+            )
+
+    # ---------- Sampling ----------#
+
+    def forward(
+        self,
+        cond,
+        deterministic=False,
+        reparameterize=False,
+    ):
+        """use both pre-trained and online policies"""
+        q1_ind, q2_ind = self.get_random_indices()
+
+        # sample an action from the BC policy
+        bc_action = self.forward_sample(
+            cond=cond,
+            deterministic=True,
+            network_override=self.bc_policy,
+        )
+
+        # sample an action from the RL policy
+        rl_action = super().forward(
+            cond=cond,
+            deterministic=deterministic,
+        )
+
+        # compute Q value of BC policy
+        q_bc_1 = self.critic_networks[q1_ind](cond, bc_action)  # (B,)
+        q_bc_2 = self.critic_networks[q2_ind](cond, bc_action)
+        q_bc = torch.min(q_bc_1, q_bc_2)
+
+        # compute Q value of RL policy
+        q_rl_1 = self.critic_networks[q1_ind](cond, rl_action)
+        q_rl_2 = self.critic_networks[q2_ind](cond, rl_action)
+        q_rl = torch.min(q_rl_1, q_rl_2)
+
+        # soft sample or greedy
+        if deterministic or not self.soft_action_sample:
+            action = torch.where(
+                (q_bc > q_rl)[:, None, None],
+                bc_action,
+                rl_action,
+            )
+        else:
+            # compute the Q weights with probability proportional to exp(\beta * Q(a))
+            qw_bc = torch.exp(q_bc * self.soft_action_sample_beta)
+            qw_rl = torch.exp(q_rl * self.soft_action_sample_beta)
+            q_weights = torch.softmax(
+                torch.stack([qw_bc, qw_rl], dim=-1),
+                dim=-1,
+            )
+
+            # sample according to the weights
+            q_indices = torch.multinomial(q_weights, 1)
+            action = torch.where(
+                (q_indices == 0)[:, None],
+                bc_action,
+                rl_action,
+            )
+        return action
+
+    # override
+    @torch.no_grad()
```

Comment:
remove, make it differentiable

Reply:
Done.
<!-- Reviewable comment -OBhfDKm4owdVJwnfzuM:b-896fix -->
<!-- Sent from Reviewable.io -->